taskfile/cmd.go

```diff
@@ -41,6 +42,18 @@ func (c *Cmd) UnmarshalYAML(unmarshal func(interface{}) error) error {
 		c.IgnoreError = cmdStruct.IgnoreError
 		return nil
 	}
+	var deferredCmd struct {
+		Defer string
+	}
+	if err := unmarshal(&deferredCmd); err == nil && deferredCmd.Defer != "" {
+		c.Defer = true
+		if strings.HasPrefix(deferredCmd.Defer, "^") {
+			c.Task = strings.TrimPrefix(deferredCmd.Defer, "^")
+		} else {
+			c.Cmd = deferredCmd.Defer
+		}
```

Comment:
This behavior is deprecated and undocumented. It was kept above for compatibility reasons (but will be eventually removed). Since here it's a new feature we should just do:

```suggestion
		c.Cmd = deferredCmd.Defer
```

Reply:
That's fine with me - but I think being able to call a task here is a use case I would like to support.

For example, I currently have a couple of projects with tasks such as `test:setup` and `test:teardown` which encompass some number of complex setup steps. During local development, I might keep the test environment set up for a prolonged period of time, and only tear down when I need to change config for a container. In CI, I'll call those tasks as a part of a `test:ci` task which sets up an environment for tests, runs them, and tears down the environment. Some projects have multiple test environments to test against different versions of external software, for example.

I could of course use `defer: "task test:setup"`. If `task` isn't in `PATH` on all machines, it gets a little more complicated, and the double-layered task output is kind of annoying (but not truly problematic).

Do you feel like this is something worth supporting? I'm envisioning something like:

```yaml
test:ci
    cmds:
        - task: test:setup
        - defer: { task: test:teardown }
        - task: test:run_unit
        - task: test:run_integration
```

 Perhaps if we want to be ambitious we could even support something like:

```yaml
defer: { task: test:teardown, vars: { dep_version: v1.2.3 } }
```

Reply:
@mccolljr Yeah, I think it makes sense to support that.

FYI, we already have a common struct for a call: https://github.com/go-task/task/blob/master/taskfile/call.go

I would consider creating a `Defer` struct (and file) in the `taskfile` package. It could look something like this:

```go
struct Defer {
  Cmd string
  Call Call
}
```

With a `UnmarshalYAML` function to handle the cmd vs. call marshalling.

In theory any of this could be valid:

```yaml
- defer: echo 'foo'
- defer: { cmd: echo 'bar' }
- defer: { task: baz }
- defer: { task: baz, vars: { FOO: bar } }
```

Let me know if you have thoughts or questions.

Reply:
@andreynering for some reason I totally missed that `Call` type.

I ended up going with a slightly modified version of the above approach to keep the code changes within the executor to a minimum, but the supported syntax is basically what you described above minus the `{ cmd: "echo 'blah'" }` version.